src/language-js/print/function.js

```diff
@@ -249,160 +248,243 @@ function printArrowFunctionSignature(path, options, print, args) {
+    (bodyDoc.label?.hug !== false &&
+      (bodyDoc.label?.embed ||
+        isTemplateOnItsOwnLine(node, options.originalText))) ||
+    node.type === "ArrowFunctionExpression" ||
+    node.type === "DoExpression"
+  );
+}
+
+/**
+ * @param {Doc[]} signatures
+ * @param {*} parent
+ * @param {Object} flags
+ * @param {boolean} flags.isCallee
+ * @param {boolean} flags.shouldBreak
+ * @param {boolean} flags.isAssignmentRhs
+ * @returns {Doc}
+ */
+function joinArrowFunctionSignatures(
   signatures,
-  shouldBreak,
-  bodyDoc,
-  tailNode
+  parent,
+  { isCallee, shouldBreak, isAssignmentRhs }
 ) {
-  const { parent, key } = path;
-  const isCallee = isCallLikeExpression(parent) && key === "callee";
-  const isAssignmentRhs = Boolean(args?.assignmentLayout);
-  const shouldPutBodyOnSeparateLine =
-    tailNode.body.type !== "BlockStatement" &&
-    !isObjectOrRecordExpression(tailNode.body) &&
-    tailNode.body.type !== "SequenceExpression";
-  const shouldBreakBeforeChain =
-    (isCallee && shouldPutBodyOnSeparateLine) ||
-    args?.assignmentLayout === "chain-tail-arrow-chain";
-
-  const groupId = Symbol("arrow-chain");
-
+  if (signatures.length === 1) {
+    return signatures[0];
+  }
   if ((isCallLikeExpression(parent) && !isCallee) || isBinaryish(parent)) {
-    signatures = [dedent(signatures[0]), ...signatures.slice(1)];
+    return group(
+      [
+        signatures[0],
+        " =>",
+        indent([line, join([" =>", line], signatures.slice(1))]),
+      ],
+      {
+        shouldBreak,
+      }
+    );
   }
-
-  // We handle sequence expressions as the body of arrows specially,
-  // so that the required parentheses end up on their own lines.
-  if (tailNode.body.type === "SequenceExpression") {
-    bodyDoc = group(["(", indent([softline, bodyDoc]), softline, ")"]);
+  if (isCallee || isAssignmentRhs) {
+    return group(join([" =>", line], signatures), {
+      shouldBreak,
+    });
   }
+  return group(indent(join([" =>", line], signatures)), {
+    shouldBreak,
+  });
+}
 
-  return group([
-    group(
-      indent([
-        isCallee || isAssignmentRhs ? softline : "",
-        group(join([" =>", line], signatures), { shouldBreak }),
-      ]),
-      { id: groupId, shouldBreak: shouldBreakBeforeChain }
-    ),
-    " =>",
-    indentIfBreak(
-      shouldPutBodyOnSeparateLine ? indent([line, bodyDoc]) : [" ", bodyDoc],
-      { groupId }
-    ),
-    isCallee ? ifBreak(softline, "", { groupId }) : "",
-  ]);
+/**
+ * @param {Doc} bodyDoc
+ * @param {Doc[]} bodyComments
+ * @param {*} parent
+ * @param {*} node
+ * @param {Object} flags
+ * @param {boolean} flags.shouldAddParensIfNotBreak
+ * @param {boolean} flags.shouldAlwaysAddParens
+ * @param {boolean} flags.shouldPutBodyOnSameLine
+ * @param {*} args
+ * @param {*} options
+ */
+function printArrowFunctionBody(
+  bodyDoc,
+  bodyComments,
+  parent,
+  node,
+  { shouldAddParensIfNotBreak, shouldAlwaysAddParens, shouldPutBodyOnSameLine },
```

Comment:
Can we pass `path` here, instead of `node`, `parent`. Maybe we can also move those `should*` check into this function?

Reply:
Passing `path` instead of `node` and `parent` looks feasible.
I'll change it later.

Moving `should*` into `printArrofFunctionBody()` is not straight forward because it also affect format of signatures.
- `shouldAddParensIfBreak` affects format of signatures.
    - `shouldAddParensIfBreak` → `shouldAddParens` → `shouldPutBodyOnSameLine` → `shouldBreakBeforeChain` → format of signatures
- `shouldAlwaysAddParens` affects format of signatures.
    - `shuoldAlwaysAddParnens` → `shouldAddParens` → `shouldPutBodyOnSameLine` → `shouldBreakBeforeChain` → format of signatures
- `shuoldPutBodyOnSameLine` → `shouldBreakBeforeChain` → format of signatures

Here, I denoted A → B when A affects B, and A → B → C when A affects B and B affects C.

We have some options. I'll write them.

Reply:
Option E.  Both  `printArrowFunction()` and `printArrowFunctionBody()` calculates `should*` independently even if they should always same.

Reply:
I see they both need those checks now.